File: 04-17-2024/Binary_Tree_Postorder_Traversal.java
//  145. Binary Tree Postorder Traversal
//
//  https://leetcode.com/problems/binary-tree-postorder-traversal/
import java.util.*;

public class Binary_Tree_Postorder_Traversal {
    public static List<Integer> ans;
    public static void solve(TreeNode root) {
        if(root == null) return;
        solve(root.left);
        solve(root.right);
        ans.add(root.val);
    }
    public static List<Integer> postorderTraversal(TreeNode root) {
        ans = new ArrayList<>();
        solve(root);
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{1,2,3,4,5,6,7};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        List<Integer> ans = postorderTraversal(root);
        System.out.println(ans.toString());
    }
}

File: 2026-02-12/Binary_Tree_Postorder_Traversal.py
from typing import Optional, List
from Tree import TreeNode, Tree


class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        cur = root

        while cur:
            if not cur.right:
                result.append(cur.val)
                cur = cur.left
            else:
                rightPointer = cur.right
                while rightPointer.left and rightPointer.left != cur:
                    rightPointer = rightPointer.left

                if not rightPointer.left:
                    result.append(cur.val)
                    rightPointer.left = cur
                    cur = cur.right
                else:
                    rightPointer.left = None
                    cur = cur.left

        result.reverse()
        return result


sol = Solution()
tree = Tree()
node = tree.array_to_tree([1,2,3,4,5,None,8,None,None,6,7,9])
print(sol.postorderTraversal(node))
